File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.cc
#include <cstdio>

int N;
int eggs[8][2];

int backtracking(int current = 0, int broken = 0) {
  if (current == N) {
    return broken;
  } else if (eggs[current][0] <= 0 || broken == N - 1) {
    return backtracking(current + 1, broken);
  } else {
    int count = 0;
    for (int next = 0; next < N; next++) {
      if (current == next || eggs[next][0] <= 0) {
        continue;
      }

      eggs[current][0] -= eggs[next][1];
      eggs[next][0] -= eggs[current][1];
      
      int ret = backtracking(
          current + 1, 
          broken + (eggs[current][0] <= 0) + (eggs[next][0] <= 0)
      );

      if (ret > count) {
        count = ret;
      }

      eggs[current][0] += eggs[next][1];
      eggs[next][0] += eggs[current][1];
    }
    return count;
  }
}

int main() {
  scanf(" %d", &N);

  for (int i = 0; i < N; i++) {
    scanf(" %d %d", &eggs[i][0], &eggs[i][1]);
  }

  printf("%d", backtracking());
  return 0;
}

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
input = open(0).readline

N = int(input())
S, W = [0] * N, [0] * N
for i in range(N):
    s, w = map(int, input().split())
    S[i], W[i] = s, w
    

def backtracking(current: int = 0, broken: int = 0) -> int:
    if broken == N - 1 or current == N:  # 모든 계란이 깨졌거나 끝까지 왔다면면
        return broken
        
    if S[current] <= 0:  # 이미 깨짐
        return backtracking(current + 1, broken)
        
    count: int = 0
    for nxt in range(N):
        if current == nxt or S[nxt] <= 0:  # 같은 계란이거나 이미 깨졌다면 skip
            continue
        
        S[current] -= W[nxt]
        S[nxt] -= W[current]
        
        count = max(count, backtracking(current + 1, broken + (S[current] <= 0) + (S[nxt] <= 0)))
        
        S[current] += W[nxt]
        S[nxt] += W[current]
    
    return count

    
print(backtracking())